### ISSUE TITLE
[Umbrella Issue] SemIf 決策模型演進與教學導向計畫：從 Zero-Shot Logits 到高校準、極低延遲的語意決策引擎

### ISSUE THREAD

**Issue by EndeavorYen** [EndeavorYen/SemIf#1]
# [Umbrella Issue] SemIf 決策模型演進計畫：從 Zero-Shot Logits 到高校準、極低延遲的語意決策引擎

> **關聯環境與硬體目標**：
> - 主要推論/實驗：**NVIDIA RTX 5080 (16GB GDDR7, Blackwell)**
> - 跨平台輕量部署：**Apple Silicon Mac mini (16GB 統一記憶體, WebGPU / MLX / MPS)**
> - 基準模型：`Qwen/Qwen3.5-4B`（SemIf 原生）、`Mapika/decider-2b`（專門訓練對照組）、TypeSafe Jev（公開報告錨點）

---

## 專案總覽與目標

[SemIf](https://github.com/TheoLeeCJ/SemIf)（原名 `openjev`）驗證了一項關鍵價值：**在不需要大模型自回歸生成字串或 JSON 的前提下，利用單次前向傳播（Single Forward Pass）讀取選項 Logits，即可實現超低延遲的型別化語意決策（Semantic If）。**

然而，目前的實作存在以下幾大工程與演算法挑戰：
1. **未校準的機率（Uncalibrated Probabilities）**：Softmax 極易過度自信，不能直接作為安全閾值（Confidence Threshold）。
2. **位置偏置（Position Bias）與字母先驗偏置（Prior Bias）**：選項顛倒會造成預測翻轉（Argmax Flip）。
3. **LM Head 全詞表計算浪費**：15 萬+ 詞表投影拖慢推論速度。
4. **缺少系統化的決策品質評測**：缺少 ECE（預期校準誤差）、Brier 分數深入分析與 OOD（超出分佈/無匹配項）檢測。

本 Umbrella Issue 旨在建立完整的改進體系，並結合 **教學驅動（Learning-Driven Development）**，在推進工程的同時，產出系統化的教學文件，幫助深入掌握：
- **LLM Logits 提取底層機制與注意力快取（KV / Prefix Cache）**
- **統計機器學習中的模型校準（Calibration）、ECE、Brier Score**
- **CUDA Graphs、受限矩陣乘法（Sliced Projection）等低延遲壓榨技巧**
- **Apple Silicon MLX 與 WebGPU 跨平台工程實踐**

---

## 階段劃分與 Sub-issues 清單

```mermaid
flowchart TD
    subgraph Track1 [第一階段：評測基準與校準理論建置]
        I1["Sub-Issue #1: 擴充校準度量工具鏈 (ECE, Brier, Flip Rate)"]
        I2["Sub-Issue #2: 接入 Mapika/decider-2b 建立對照基準"]
    end

    subgraph Track2 [第二階段：零訓練推論期後處理校準]
        I3["Sub-Issue #3: 實作溫度縮放 (Temperature Scaling) 與空先驗去偏"]
        I4["Sub-Issue #4: 實作前綴快取選項輪換 (Permutation Ensembling)"]
    end

    subgraph Track3 [第三階段：決策邊界健全化與 OOD 防護]
        I5["Sub-Issue #5: 獨立 Sigmoid 門控與自由能 OOD 檢測"]
    end

    subgraph Track4 [第四階段：RTX 5080 極限加速]
        I6["Sub-Issue #6: 受限投影 (Sliced LM Head) 消除冗餘計算"]
        I7["Sub-Issue #7: CUDA Graphs 與 torch.compile 4ms 延遲挑戰"]
    end

    subgraph Track5 [第五階段：Mac mini 跨平台移植]
        I8["Sub-Issue #8: Apple Silicon (MLX / MPS) 原生推理引擎移植"]
    end

    Track1 --> Track2 --> Track3 --> Track4 --> Track5
```

## 執行狀態追蹤 (Tracked Sub-Issues)
- [x] #2 [Sub-Issue 1] 擴充校準度量工具鏈 (ECE, Brier 分數, 排列翻轉率)
- [x] #3 [Sub-Issue 2] 評測環境整合 Mapika/decider-2b 對照基準
- [x] #4 [Sub-Issue 3] 推論期零訓練去偏與溫度校準 (Temperature Scaling & Prior Calibration)
- [x] #5 [Sub-Issue 4] 選項順序輪換集成 (Permutation Ensembling via Prefix Reuse)
- [x] #6 [Sub-Issue 5] 決策邊界健全化 (Sigmoid 獨立門控與自由能 OOD 拒絕)
- [x] #7 [Sub-Issue 6] RTX 5080 硬體適配與受限投影 (Restricted LM Head)
- [x] #8 [Sub-Issue 7] CUDA Graphs 與 torch.compile 極限延遲挑戰 (衝擊 4~8ms)
- [x] #9 [Sub-Issue 8] 拓展至 Apple Silicon Mac mini (16GB 統一記憶體)

---

### [Sub-Issue 1] #2 評測指標工具鏈擴充（ECE、Brier 分數、排列翻轉率）
- **核心目標**：為語意決策建立嚴謹的統計評測指標，不再只看「準確率（Accuracy）」，更能衡量「模型說 80% 信心時是否真有 80% 勝率」。
- **實作重點**：
  - 在 `benchmarks/evaluate.py` 中新增 `compute_ece(probabilities, labels, n_bins=15)`。
  - 在 `benchmarks/evaluate_perturbations.py` 整合選項對調（Option Reversal）的翻轉率統計。
  - 增加 Brier Score 與可靠度圖（Reliability Diagram）數據導出。
- **隨附教學文件**：
  - `docs/tutorials/01_model_calibration_and_ece.md`：
    - 什麼是模型校準？為什麼傳統 LLM 的 Softmax 會「過度自信」？
    - ECE 與 Brier Score 的數學原理與直觀解釋。
    - 如何閱讀可靠度圖（Reliability Diagram）並識別過度自信或平緩分佈。

---

### Sub-Issue #2: 評測環境整合 `Mapika/decider-2b` 對照基準
- **核心目標**：將專門微調決策任務的 `Mapika/decider-2b` 作為黃金對照組（Baseline），在 SemIf 既有的資料集上執行對比評測。
- **實作重點**：
  - 撰寫適配器（Adapter）使 `Mapika/decider-2b` 能讀取 SemIf 的 JSONL 格式。
  - 在 `manifests/authored_decisions.jsonl`、`manifests/typesafe_eval.jsonl` 與 `wanli.jsonl` 上運行測試。
  - 橫向對比：`Qwen3.5-4B Direct` vs `Mapika/decider-2b` vs `TypeSafe Jev`。
- **隨附教學文件**：
  - `docs/tutorials/02_decision_native_vs_causal_lm.md`：
    - 通用生成式因果語言模型（Causal LM）與專門決策模型（Decision-Native Model）的差異。
    - `decider-2b` 的訓練原理：如何用 Proper Scoring Rule 訓練 System One 決策。

---

### Sub-Issue #3: 推論期零訓練去偏與溫度校準（Temperature Scaling & Prior Calibration）
- **核心目標**：不改動任何模型權重，純靠推論計算消除 70% 以上的字母偏好與極端機率。
- **實作重點**：
  - 修改 `core.py` 中的 `softmax` 函式，引入可配置的溫度係數 $T$。
  - 建立 `calibrate_temperature.py`，使用 SciPy 擬合使驗證集 NLL / ECE 最小化的最佳標量 $T$。
  - 實作空上下文去偏（Context-Free Prior Calibration）：預先計算虛擬輸入（`state="N/A"`）下的無條件先驗 $z_{\text{null}}$，並在推論時執行 $z_{\text{calib}} = z_{\text{raw}} - z_{\text{null}}$。
- **隨附教學文件**：
  - `docs/tutorials/03_inference_time_calibration.md`：
    - 深入解析溫度縮放（Temperature Scaling）的數學本質與凹凸性證明。
    - 空上下文校準（Calibrate Before Use）原理：為何 LLM 對字母 'A' 天生偏愛？

---

### Sub-Issue #4: 選項順序輪換集成（Permutation Ensembling via Prefix Reuse）
- **核心目標**：徹底消除 Prompt 中選項先後順序帶來的「首因效應」與「近因效應」，將選項倒轉翻轉率壓至接近 0。
- **實作重點**：
  - 實作選項置換評估器：若輸入選項為 $[A, B]$，同時評估 $[A, B]$ 與 $[B, A]$。
  - **極低延遲保證**：利用 `shared.py` / `serial.py` 的 KV Cache 機制，狀態（State）部分僅計算一次，僅需對後綴的 2~3 種排列進行極短的 Forward 計算。
  - 對齊各選項 ID 後取幾何/算術平均。
- **隨附教學文件**：
  - `docs/tutorials/04_prefix_cache_and_position_bias.md`：
    - Transformer 的位置編碼（RoPE）如何引發 Position Bias？
    - KV Cache / Prefix Caching 的記憶體架構與工作機制圖解。

---

### Sub-Issue #5: 決策邊界健全化（Sigmoid 獨立門控與自由能 OOD 拒絕）
- **核心目標**：打破 Softmax 封閉世界假設，使系統具備「拒絕回答（None of the above）」與多條件匹配能力。
- **實作重點**：
  - 實作獨立 Sigmoid 分類模式：$p_i = \sigma(z_i)$。
  - 引入自由能門控（Energy-based Gating）：$E(x) = -T \log \sum_i e^{z_i/T}$，用於辨識缺失證據（Missing Evidence）或異常輸入。
- **隨附教學文件**：
  - `docs/tutorials/05_open_world_and_ood_detection.md`：
    - Closed-World（Softmax）與 Open-World（Sigmoid / Energy）的架構思維差異。
    - 自由能（Free Energy）如何充當 Agent 工作流中的安全護欄（Safety Guardrail）。

---

### Sub-Issue #6: RTX 5080 硬體適配與受限投影（Restricted LM Head）
- **核心目標**：在 RTX 5080（Blackwell 架構）上消除 Qwen3.5-4B 在 15 萬詞表上的巨量無效矩陣乘法。
- **實作重點**：
  - 修改 `direct.py` 的 `_forward()`：不呼叫 `model.forward()` 輸出全詞表 Logits，而是提取最後一層的 Hidden State，透過 Sliced Linear Weight 僅投影需要的 $K$ 個選項 Token（$K \le 16$）。
  - 驗證運算量從 $O(d_{\text{hidden}} \times 151936)$ 降至 $O(d_{\text{hidden}} \times K)$（減少 99.9% 乘加運算）。
- **隨附教學文件**：
  - `docs/tutorials/06_transformer_lm_head_optimization.md`：
    - 解構 Hugging Face Causal LM 的最後一層（`lm_head`）矩陣計算。
    - 如何用 PyTorch 進行權重切片（Weight Slicing），省下數百 MB 的記憶體頻寬。

---

### Sub-Issue #7: CUDA Graphs 與 torch.compile 極限延遲挑戰（衝擊 4~8ms）
- **核心目標**：消除 Python 直譯器開銷與 CUDA Kernel 發射延遲，挑戰單次決策 4~8ms。
- **實作重點**：
  - 參考 `decider` 的 Shape-Bucket 靜態圖捕獲機制（CUDA Graphs）。
  - 配合 `torch.compile(mode="reduce-overhead")`，固定維度的後綴前向運算。
  - 實測 RTX 5080 上的 End-to-End P50 延遲與並發吞吐量。
- **隨附教學文件**：
  - `docs/tutorials/07_cuda_graphs_and_compilation.md`：
    - 什麼是 CPU Overhead 與 Kernel Launch Latency？
    - CUDA Graphs 捕獲與重播原理教學。

---

### Sub-Issue #8: 拓展至 Apple Silicon Mac mini（16GB 統一記憶體）
- **核心目標**：實現不依賴高階 NVIDIA 顯卡、在 Mac mini 上流暢運行的本地語意決策引擎。
- **實作重點**：
  - **路徑 A（原生 MLX 移植）**：利用 Apple 官方的 `mlx` / `mlx-lm` 庫，編寫極簡的單次 Forward + Logits 讀取腳本，享受零拷貝（Zero-Copy）統一記憶體加速。
  - **路徑 B（PyTorch MPS 後端相容）**：修改裝置指派邏輯，支援 `mps` 裝置加速。
  - **路徑 C（靜態 WebGPU 演示驗證）**：驗證既有的 `webgpu-demo/` 在 Safari / Chrome 上的 2B/4B 表現。
- **隨附教學文件**：
  - `docs/tutorials/08_mac_silicon_and_mlx_deployment.md`：
    - Apple Silicon 統一記憶體架構（UMA）與 NVIDIA 顯存架構的差異。
    - MLX 框架入門：如何用 MLX 快速抓取特定 Token 的 Logits。

---

## 交付原則與規範（依據 AGENTS.md）

1. **嚴守可重現性**：
   - 不修改既有基準檔案 `results/phase1-summary.json`，所有新實驗與優化數據產出至獨立路徑（如 `results/phase2/...`），採 Create-Only 模式。
   - 所有改動必須通過 `pytest -q` 與現有的 Checksum 驗證。
2. **教學與代碼並重**：
   - 每完成一個 Sub-issue，必須同步產出對應的 Markdown 教學手冊於 `docs/tutorials/`，內含理論背景、代碼註解、前後數據對比與直觀圖表。

**Comment by EndeavorYen** [comment]
🎉 第一階段所有 8 個子任務（Sub-Issue 1 至 8）全數完工並通過驗證測試，PR #10 ~ #17 已全部合併至 master。結案！
